### PR TITLE
Hand-code remote-write protobuf

### PR DESCRIPTION
**Pull request by bboreham**
Draft, as you may not want to add the complexity, but I thought it was worth seeing the performance achievable.

```
name          old time/op    new time/op    delta
SampleSend-4    1.70ms ± 0%    1.12ms ± 4%  -34.44%  (p=0.016 n=4+5)

name          old alloc/op   new alloc/op   delta
SampleSend-4    47.4kB ± 0%    29.6kB ± 1%  -37.57%  (p=0.008 n=5+5)

name          old allocs/op  new allocs/op  delta
SampleSend-4       140 ± 0%       110 ± 0%  -21.43%  (p=0.008 n=5+5)
```

The code is pretty much as generated by the proto compiler, but working off the data structure created in #9934. We benefit from not creating the intermediate structs that protoc normally uses. 
As an additional step I re-wrote some of the varint code to go directly to the number of 7-bit words required, and optimise for small numbers

Another step I experimented with was merging external labels at the point of marshalling rather than creating an additional slice for each series in `processExternalLabels` called from `StoreSeries`. This does reduce heap size but increases CPU about 15% on every send, so I left it out.

### CODE REVIEW

File: storage/remote/marshal.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
+// Copyright 2013 The Prometheus Authors
```
Comment (colega) [inline storage/remote/marshal.go:1]: I guess you need to update the date here.

**Comment by bboreham** [comment]
@csmarchbanks do you have a view on this?

**Review by csmarchbanks — COMMENTED** [review]
I can give this some further review after break. Generally I think I would be OK with the more complex custom code, especially since any issues should be caught be tests. It should be fairly easy to remove as well if it does cause maintenance issues.

@cstyan what do you think as well?

**Comment by cstyan** [comment]
I personally wouldn't want to maintain the marshalling code, so it's up to Chris (or if Bryan is offering to keep with maintenance of the marshalling code). Given that this version of the protocol is pretty stable and likely won't change until a new protocol version I think we're fine to move forward, the improvements shown in the benchmark are pretty compelling.

**Comment by bboreham** [comment]
>  if Bryan is offering to keep with maintenance of the marshalling code

Sure, I can do that.  (Assuming there is some value in a random person on the internet committing to do something)

**Comment by roidelapluie** [comment]
I also think we could move forward here.

File: storage/remote/marshal.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+		copy(data[i:], m.Name)
+		i = encodeSize(data, i, len(m.Name))
+		i--
+		data[i] = labelNameTag
+	}
+	return len(data) - i, nil
+}
+
+func sizeVarint(x uint64) (n int) {
+	// Most common case first
```
Comment (bwplotka) [inline storage/remote/marshal.go:155]:
```suggestion
	// Most common case first.
```

File: storage/remote/marshal.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+	for v >= 1<<7 {
+		data[offset] = uint8(v&0x7f | 0x80)
+		v >>= 7
+		offset++
+	}
+	data[offset] = uint8(v)
+	return base
+}
+
+// Special code for the common case that a size is less than 128
```
Comment (bwplotka) [inline storage/remote/marshal.go:188]:
```suggestion
// Special code for the common case that a size is less than 128.
```

File: storage/remote/marshal.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+		offset--
+		data[offset] = uint8(v)
+		return offset
+	}
+	return encodeVarint(data, offset, uint64(v))
+}
+
+func sampleOrExemplarSliceSize(data []sampleOrExemplar) (n int) {
+	for _, e := range data {
+		l := e.protoSize()             // for the sample or exemplar struct
```
Comment (bwplotka) [inline storage/remote/marshal.go:200]:
```suggestion
		l := e.protoSize()             // For the sample or exemplar struct.
```

File: storage/remote/marshal.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+		l += 1 + sizeVarint(uint64(l)) // for the timeSeries struct that wraps it
```
Comment (bwplotka) [inline storage/remote/marshal.go:201]:
```suggestion
		l += 1 + sizeVarint(uint64(l)) // For the timeSeries struct that wraps it.
```

File: storage/remote/queue_manager.go [hunk 6]
```diff
@@ -1198,48 +1196,24 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue *queue) {
 			}
 			s.mtx.Unlock()
 			if len(batch) > 0 {
-				nPendingSamples, nPendingExemplars := s.populateTimeSeries(batch, pendingData)
```
Comment (bwplotka) [inline storage/remote/queue_manager.go:1201]: Hm, those pending samples and pending exemplars changes are not related to marshalling are they?

1. Are we sure we don't need this functionality anymore?
2. Are we sure when the removal this functionality is NOT what is faster/more efficient vs marshalling changes?

File: storage/remote/codec_test.go [hunk 7]
```diff
@@ -336,7 +336,7 @@ func TestMetricTypeToMetricTypeProto(t *testing.T) {
 }
 
 func TestDecodeWriteRequest(t *testing.T) {
```
Comment (bwplotka) [inline storage/remote/codec_test.go:338]: Can we test building request with proto TimeSeries AND populating samples (as it was) and marshal the same data using our marshalling and compare binary (I think we expect it to be 1:1?).

**Review by bwplotka — COMMENTED** [review]
Nice, we can move with it if the results are true, but I wonder if we truly isolate changes when benchmarking 🤔 

It would be nice to split this PR if possible to marshalling change vs refactor so reverting to generated protobuf is straightforward if needed.

File: storage/remote/queue_manager.go (hunk 6 above)
Comment (bboreham) [reply storage/remote/queue_manager.go:1201]: Yes it is related.  The code now generates protobuf messages directly from Prometheus data structures, and does not convert to protobuf structs.  The function `populateTimeSeries` used to do this conversion, and has been removed.

You're right that this is a large part of making it faster.  This is stated in the PR description and in the git commit message.

**Comment by roidelapluie** [comment]
We have discussed this during our bug scrub and decided with @bboreham to close this pull request for now.
